### DISCUSSION TITLE
Native support for unbound numeric computations / BigInt / AIR model / Arbitrary wide integers

### DISCUSSION THREAD

**Discussion by DmT021** [hylo-lang/hylo#892] [Language design]
The only case where I still use python instead of Swift scripting is when I need to compute some math involving arbitrary large numbers. 
Fixed width integer operations combines badly with natural math. I.e. the result of multiplication of two 32 bit integers will be 64 bit integer. If we try to model math operations the way they produce fixed width integers with width enough to guarantee safety, we will run out of space immediately. 
There are several approaches to this:
- BigInt-like structures. The underlying storage grows dynamically while computing (like `Array`).
- Fixed width integers, but with arbitrary width specified by programmer (`WideInt<512>` - means integer of size 512 bits, like `Buffer`)

An interesting aspect of having such types as natives is the ability to provide transparent vectorization for them.

**Comment by kyouko-taiga** [comment]
I'm planning on adding `WideInt<n>` to the standard library once I'll have tackled generic value parameters:

```
type WideInt<n: Int> {
  var value: Builtin.i<n> // <- this is LLVM `in` (e.g., `i128`, `i256`, ...)
  ...
}
```
Ditto for `WideUInt<n>`.

I think `BigInt` handles different yet equally valid use cases.
